### PR TITLE
#102 refactored user class

### PR DESCRIPTION
**Pull request by ammaratef45**
PR for #102, #50 
refactoring to get rid of firebase database and some of its logic in the code.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#103](https://github.com/ammaratef45/attendance/issues/103) is now in scope, role is `REV`

**Comment by 0crat** [comment]
@ammaratef45[/z](https://www.0crat.com/u/ammaratef45) everybody who has role `REV` is banned at [#103](https://github.com/ammaratef45/attendance/issues/103); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

**Comment by Bahaaib** [comment]
@rultor merge

File: attendance/test/api_test.dart [hunk 1]
```diff
@@ -121,7 +121,7 @@ void main() {
         );
       }
     });
-  });
+  }, skip: true);
 
   group('test getting info', () {
     final Future<String> t = token();
```
Comment (Bahaaib) [inline attendance/test/api_test.dart]: @ammaratef45 naming too

File: attendance/test/api_test.dart [hunk 2]
```diff
@@ -168,4 +168,54 @@ void main() {
       }
     });
   });
+
+  group('test getting attendance', () {
+    final Future<String> t = token();
+    const String _attendanceKey = '-LXDFDibiC8EyKRg96Zv';
```
Comment (Bahaaib) [inline attendance/test/api_test.dart]: @ammaratef45 Isn't it better if we created a generator class that bind these values without hardcoding them?

File: attendance/test/attendance_test.dart [hunk 3]
```diff
@@ -2,21 +2,55 @@ import 'package:attendance/backend/attendance.dart';
...
+  final Map<String, dynamic> _attendanceData = <String, dynamic>{
     'key' : '-LXDFDibiC8EyKRg96Zv',
     'sessionDate' : '2019-01-27T09:40:45.883409',
     'arriveDate' : '2019-01-27T09:40:45.883409',
     'leaveDate' : '2019-01-27T09:40:45.883409',
     'className' : 'BIOCHEMICAL'
   };
-  final Attendance attendance = Attendance.fromMap(attendanceData);
+  final Attendance _attendance = Attendance.fromMap(_attendanceData);
+  final Map<String, dynamic> _attendanceNotLeavedData = <String, dynamic>{}
```
Comment (Bahaaib) [inline attendance/test/attendance_test.dart]: @ammaratef45 typos!

File: attendance/test/attendance_test.dart [hunk 4]
```diff
@@ -2,21 +2,55 @@ import 'package:attendance/backend/attendance.dart';
...
+      expect(att.className, null);
+      expect(att.session.date, null);
+      expect(att.arriveDate, null);
+      expect(att.leaveDate, null);
+      expect(att.key, _attendanceData['key']);
+    });
+  }, skip: true);
+
+  group('leaving', () {
+    test('is leaved', () {
```
Comment (Bahaaib) [inline attendance/test/attendance_test.dart]: @ammaratef45 typos!

File: attendance/test/attendance_test.dart [hunk 5]
```diff
@@ -2,21 +2,55 @@ import 'package:attendance/backend/attendance.dart';
...
+      expect(_notLeavedAttendance.isLeaved(), false);
+      expect(_attendance.isLeaved(), true);
+    });
+    test('leaving working', () {
+      _notLeavedAttendance.leave('2019-01-27T09:40:45.883409');
+      expect(_notLeavedAttendance.leaveDate, '2019-01-27T09:40:45.883409');
+      expect(_notLeavedAttendance.isLeaved(), true);
```
Comment (Bahaaib) [inline attendance/test/attendance_test.dart]: @ammaratef45 Typos. Typos everywhere!

File: attendance/test/api_test.dart [hunk 6]
```diff
@@ -38,18 +38,18 @@ void main() {
 
   group('session calls', () {
     final Future<String> t = token();
```
[the remark this answers is not in the document]
Comment (ammaratef45) [reply attendance/test/api_test.dart]: @Bahaaib I agree, I will edit it

File: attendance/test/api_test.dart [hunk 7]
```diff
@@ -63,7 +63,7 @@ void main() {
     });
     test('leave fails due to wrong token', () async {
```
[the remark this answers is not in the document]
Comment (ammaratef45) [reply attendance/test/api_test.dart]: @Bahaaib what else you suggest?

File: attendance/test/api_test.dart (hunk 2 above)
Comment (ammaratef45) [reply attendance/test/api_test.dart]: @Bahaaib I'll leave a puzzle for it

File: attendance/test/attendance_test.dart (hunk 3 above)
Comment (ammaratef45) [reply attendance/test/attendance_test.dart]: @Bahaaib please explain what is wrong

File: attendance/test/attendance_test.dart (hunk 4 above)
Comment (ammaratef45) [reply attendance/test/attendance_test.dart]: @Bahaaib please explain what is wrong

File: attendance/test/attendance_test.dart (hunk 5 above)
Comment (ammaratef45) [reply attendance/test/attendance_test.dart]: @Bahaaib please explain what is wrong

File: attendance/test/api_test.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (Bahaaib) [reply attendance/test/api_test.dart]: @ammaratef45 if you meant the noun of "to leave" then it's departure

File: attendance/test/attendance_test.dart (hunk 3 above)
Comment (Bahaaib) [reply attendance/test/attendance_test.dart]: @ammaratef45 left NOT leaved

**Comment by ammaratef45** [comment]
@Bahaaib I've solved the army of the typos, please look again.

**Comment by 0crat** [comment]
This pull request [#103](https://github.com/ammaratef45/attendance/issues/103) is assigned to @aqeel-hashim[/z](https://www.0crat.com/u/aqeel-hashim), here is [why](https://www.0crat.com/footprint/CH951175M/b4809e4f-cafc-4424-9d27-465429615e7f); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @ammaratef45[/z](https://www.0crat.com/u/ammaratef45) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by 0crat** [comment]
This pull request [#103](https://github.com/ammaratef45/attendance/issues/103) is assigned to @bahaaib[/z](https://www.0crat.com/u/bahaaib), here is [why](https://www.0crat.com/footprint/CH951175M/3736bc47-1922-4d79-87ac-f49f43bb4993); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @ammaratef45[/z](https://www.0crat.com/u/ammaratef45) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; @bahaaib[/z](https://www.0crat.com/u/bahaaib) is not a member of this project yet, but they can request to join, as [§1](http://www.zerocracy.com/policy.html#1) explains; there will be no monetary reward for this job

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @ammaratef45[/z](https://www.0crat.com/u/ammaratef45)

**Comment by ammaratef45** [comment]
@rultor merge it

**Comment by Bahaaib** [comment]
@ammaratef45 I've already burnt it..

**Comment by 0crat** [comment]
Quality is low, no payment, see [§31](http://www.zerocracy.com/policy.html#31):

**Comment by 0crat** [comment]
Job was finished in 22 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
The job [#103](https://github.com/ammaratef45/attendance/issues/103) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @ammaratef45[/z](https://www.0crat.com/u/ammaratef45)

**Pull request merged**